river/compat/pytorch.py

```diff
@@ -1,53 +1,231 @@
+import collections
+import inspect
+import typing
+
 import torch
 
 from .. import base
 
-__all__ = ["PyTorch2RiverRegressor"]
+__all__ = ["PyTorch2RiverBase", "PyTorch2RiverRegressor", "PyTorch2RiverClassifier"]
+
 
+class PyTorch2RiverBase(base.Estimator):
+    """An estimator that integrates neural Networks from PyTorch."""
 
-class PyTorch2RiverBase:
-    def __init__(self, net, loss_fn, optimizer, batch_size, x_tensor, y_tensor):
-        self.net = net
+    def __init__(
+        self,
+        build_fn,
+        loss_fn: typing.Type[torch.nn.modules.loss._Loss],
+        optimizer_fn: typing.Type[torch.optim.Optimizer] = torch.optim.Adam,
+        learning_rate=1e-3,
+        seed=42,
+        **net_params,
+    ):
+        self.build_fn = build_fn
         self.loss_fn = loss_fn
-        self.optimizer = optimizer
-        self.batch_size = batch_size
-        self.x_tensor = x_tensor
-        self.y_tensor = y_tensor
-        self._x_batch = [None] * batch_size
-        self._y_batch = [None] * batch_size
-        self._batch_i = 0
+        self.loss = loss_fn()
+        self.optimizer_fn = optimizer_fn
+        self.learning_rate = learning_rate
+        self.net_params = net_params
+        self.seed = seed
+        torch.manual_seed(seed)
+        self.net = None
+
+    def _unit_test_params(self):
+        def build_torch_linear_regressor(n_features):
+            net = torch.nn.Sequential(torch.nn.Linear(n_features, 1))
+            return net
+
+        return {
+            "build_fn": build_torch_linear_regressor,
+            "loss_fn": torch.nn.MSELoss,
+            "optimizer_fn": torch.optim.SGD,
+        }
+
+    def _learn_one(self, x: torch.Tensor, y: torch.Tensor):
+        self.net.zero_grad()
+        y_pred = self.net(x)
+        loss = self.loss(y_pred, y)
+        loss.backward()
+        self.optimizer.step()
 
     def learn_one(self, x, y):
+        if self.net is None:
+            self._init_net(n_features=len(list(x.values())))
+
+        x = torch.Tensor([list(x.values())])
+        y = torch.Tensor([[y]])
+        self._learn_one(x=x, y=y)
+        return self
```

Comment:
I agree that it makes sense to remove this fake mini-batching in `*_one` methods.

However, I think we should also have `learn_many` and `predict(_proba)_many` methods. What do you think?

Reply:
At least it would boost the tensor conversion, I'll add that.